File: benchmarks/cpp/file operations/file_read.cpp
#include <iostream>
#include <fstream>
#include <chrono>
using namespace std;
using namespace chrono;

int main() {
    const char* filename = "testfile.txt";
    volatile size_t bytes_read = 0;

    auto start = high_resolution_clock::now();

    ifstream file(filename, ios::binary);
    char buffer[4096];

    while (file.read(buffer, sizeof(buffer))) {
        bytes_read += file.gcount();
    }
    bytes_read += file.gcount(); // last partial read

    auto end = high_resolution_clock::now();
    duration<double> elapsed = end - start;

    cout << "C++ File Read Time: " << elapsed.count() << " seconds\n";
    return 0;
}

File: benchmarks/rust/file operations/file_read.rs
use std::fs::File;
use std::io::Read;
use std::time::Instant;

fn main() {
    let mut file = File::open("testfile.txt").expect("File not found");
    let mut buffer = [0u8; 4096];
    let mut _bytes_read: usize = 0;

    let start = Instant::now();

    loop {
        let n = file.read(&mut buffer).unwrap();
        if n == 0 {
            break;
        }
        _bytes_read += n;
    }

    let elapsed = start.elapsed();
    println!("Rust File Read Time: {:.6} seconds", elapsed.as_secs_f64());
}
